extensions/s3/src/main/java/io/deephaven/extensions/s3/S3CompletableOutputStream.java

```diff
@@ -268,61 +313,102 @@ private void sendPartRequest(final OutgoingRequest request) throws IOException {
                 .partNumber(nextPartNumber)
                 .build();
         request.buffer.flip();
-        request.future = s3AsyncClient.uploadPart(uploadPartRequest,
-                AsyncRequestBody.fromByteBufferUnsafe(request.buffer));
         request.partNumber = nextPartNumber;
+        request.sdkUploadFuture = s3AsyncClient.uploadPart(uploadPartRequest,
+                AsyncRequestBody.fromByteBufferUnsafe(request.buffer))
+                .whenComplete((uploadPartResponse, throwable) -> {
+                    try {
+                        if (throwable == null) {
+                            completedParts.add(
+                                    SdkPojoConversionUtils.toCompletedPart(uploadPartResponse, request.partNumber));
+                        } else {
+                            failAll(new IOException("Failed to upload part " + request.partNumber + " for uri " + uri,
+                                    throwable));
+                        }
+                    } finally {
+                        request.releaseBuffer();
```

Comment:
It's possible that we have created bufferedPartRequest, but get an error that stops of from reaching this point of code, in which case we've taken from the pool but never released it, leading to a leak.